File: min-max-heap/MinMaxHeap.js
class MinHeap {
    constructor(array) {
      this.heap = this.buildHeap(array);
    }
  
    buildHeap(array) {
      let firstParentIdx = Math.floor((array.length - 2) / 2);
      for(let currentIdx = firstParentIdx; currentIdx >= 0; currentIdx--) {
        this.siftDown(currentIdx, array.length - 1, array);
      }
      return array;
    }
  
    siftDown(currentIdx, endIdx, heap) {
      let childOneIdx = (currentIdx * 2) + 1;
      while(childOneIdx <= endIdx) {
        let childTwoIdx = (currentIdx * 2) + 2 <= endIdx ? (currentIdx * 2) + 2 : -1;
        let idxToSwap = childOneIdx;
        if(childTwoIdx != -1 && heap[childTwoIdx] < heap[childOneIdx]){
            idxToSwap = childTwoIdx;
        }
        if(heap[idxToSwap] < heap[currentIdx]){
            this.swapAt(idxToSwap, currentIdx, heap);
            currentIdx = idxToSwap;
            childOneIdx = (currentIdx * 2) + 1;
            continue;
        }
        break;
      }
    }
  
    siftUp(currentIdx, heap) {
      let parentIdx = Math.floor((currentIdx - 1) / 2);
      while(currentIdx >= 1 && heap[currentIdx] < heap[parentIdx]) {
        this.swapAt(parentIdx, currentIdx, heap);
        currentIdx = parentIdx;
        parentIdx = Math.floor((currentIdx - 1) / 2);
      }
    }
  
    peek() {
      return this.heap[0];
    }
  
    remove() {
      this.swapAt(0, this.heap.length - 1, this.heap);
      const itemToRemove = this.heap.pop();
      this.siftDown(0, this.heap.length - 1, this.heap);
      return itemToRemove;
    }
  
    insert(value) {
      this.heap.push(value);
      this.siftUp(this.heap.length - 1, this.heap);
    }
    swapAt(i, j, heap) {
        [heap[i], heap[j]] = [heap[j], heap[i]];
    }
}
  
exports.MinHeap = MinHeap;

File: min-max-heap/MinMaxHeap.cs
using System;
using System.Collections.Generic;
public class Program {
	public class MinMaxHeap {
		public List<int> heap = new List<int>();

		public MinHeap(List<int> array) {
			heap = buildHeap(array);
		}

		public List<int> buildHeap(List<int> array) {
			int firstParentIdx = (array.Count - 2) / 2;
            for(int i = firstParentIdx; i >= 0; i--) {
                siftDown(i, array.Count - 1, array);
            }
            return array;
		}

		public void siftDown(int currentIdx, int endIdx, List<int> heap) {
			int childOneIdx = (currentIdx * 2) + 1;
            while(childOneIdx <= endIdx) {
                int childTwoIdx = (currentIdx * 2) + 2 <= endIdx ? (currentIdx * 2) + 2 : -1;
                int idxToSwap = childOneIdx;
                if(childTwoIdx != -1 && heap[childTwoIdx] < heap[childOneIdx]) {
                    idxToSwap = childTwoIdx;
                }
                if(heap[idxToSwap] < heap[currentIdx]) {
                    swapAt(idxToSwap, currentIdx, heap);
                    currentIdx = idxToSwap;
                    childOneIdx = (currentIdx * 2) + 1;
                    continue;
                } 
                return;
            }
		}

		public void siftUp(int currentIdx, List<int> heap) {
			int parentIdx = (currentIdx - 1) / 2;
            while(currentIdx >= 1 && heap[currentIdx] < heap[parentIdx]) {
                swapAt(currentIdx, parentIdx, heap);
                currentIdx = parentIdx;
                parentIdx = (currentIdx - 1) / 2;
            }
		}

		public int Peek() {
    		return heap[0];
		}

		public int Remove() {
            swapAt(0, heap.Count - 1, heap);
            int itemToRemove = heap[heap.Count - 1];
            heap.RemoveAt(heap.Count - 1);
            siftDown(0, heap.Count - 1, heap);
            return itemToRemove;
		}

		public void Insert(int value) {
			heap.Add(value);
            siftUp(heap.Count - 1, heap);
		}

        public void swapAt(int i, int j, List<int> heap) {
            int temp = heap[i];
            heap[i] = heap[j];
            heap[j] = temp;
        }
	}
}
